src/TimeRestricted.sol

```diff
@@ -131,70 +267,254 @@ contract TimeRestricted is BaseGuard {
+        /// store number of modules in transient storage
+        _traverseModules(
+            SENTINEL_MODULES,
+            tstoreValueModule,
+            tstoreValueModule
+        );
+    }
+
+    /// @notice stores an address for a module in the module mapping
+    /// @param slot address to store the value in
+    /// @param value value to store in the slot
+    function tstoreValueModule(uint256 slot, uint256 value) private {
+        uint256 calculatedSlot = uint256(
+            keccak256(abi.encode(slot, MODULE_TSTORE_OFFSET))
+        );
+        assembly {
+            tstore(calculatedSlot, value)
+        }
+    }
+
+    /// @notice checks whether an address for a module is stored in
+    /// the module mapping
+    /// @param slot address to check the value in
+    /// @param value expected in the given slot
+    function checktTstoreValueModule(
+        uint256 slot,
+        uint256 value
+    ) private view {
+        uint256 calculatedSlot = uint256(
+            keccak256(abi.encode(slot, MODULE_TSTORE_OFFSET))
+        );
+        uint256 storedValue;
+
+        assembly {
+            storedValue := tload(calculatedSlot)
+        }
+
+        require(storedValue == value, "TimeRestricted: value mismatch");
+    }
+
+    /// @notice recursively traverse through all modules by asking the Safe for its
+    /// modules in a paginated manner, storing the module addresses in TSTORE slots
+    /// and finally storing the total number of module addresses in a TSTORE slot.
+    /// Algorithm:
+    ///     query for 10 modules at a time, starting at sentinel address
+    ///     if the number of modules is less than 10, store the number of modules and total number of module addresses
+    ///     else, recurse with the next address and the total number of modules found so far
+    function _traverseModules(
+        address start,
+        function(uint256, uint256) internal moduleOperation,
+        function(uint256, uint256) internal moduleLengthOperation
+    ) private returns (uint256 moduleAmountFound) {
+        (address[] memory modules, address next) = Safe(payable(msg.sender))
+            .getModulesPaginated(start, PAGE_SIZE);
+        uint256 moduleLength = modules.length;
+
+        for (uint256 i = 0; i < moduleLength; i++) {
+            uint256 moduleAddress = uint256(uint160(modules[i]));
+            // there should be no overlap between modules and owners so this operation is safe
+            // store modules in TSTORE slots
+            // store the number of modules in another TSTORE slot
+            moduleOperation(moduleAddress, 1);
+        }
+
+        /// if next == modules[modules.length - 1], we need to recurse
+        /// otherwise we are at the end of the modules list
+
+        /// if there are less than page_size modules, or the next is
+        /// the sentinel module, traversal ends
+        if (modules.length < PAGE_SIZE || next == SENTINEL_MODULES) {
+            moduleLengthOperation(
```

Comment:
I think this will store wrong length if number of modules is more than page size. Eg: If there are 15 modules it will store 5 in the tstore.